crates/matrix-sdk/src/encryption/backups/mod.rs

```diff
@@ -0,0 +1,1157 @@
+    ///
+    /// * As a `m.secret.send` to-device event from a trusted device
+    /// * From the `m.megolm_backup.v1` global account data event.
+    ///
+    /// In both cases the method will compare the currently active backup
+    /// version to the backup recovery key and if we have received a
+    /// machting key activate backups on this device and start uploading
+    /// room keys to the backup.
+    ///
+    /// Returns true if backups are already or have been enabled, otherwise
+    /// false.
+    #[instrument(skip_all)]
+    pub(crate) async fn maybe_enable_backups(
+        &self,
+        maybe_recovery_key: &str,
+    ) -> Result<bool, Error> {
+        let _guard = self.client.locks().backup_modify_lock.lock().await;
+
+        // Create a future here which allows us to catch any failure that might happen
+        // so we can later on fall back to the correct `BackupState`.
+        let future = async {
+            self.set_state(BackupState::Enabling);
+
+            let olm_machine = self.client.olm_machine().await;
+            let olm_machine = olm_machine.as_ref().ok_or(Error::NoOlmMachine)?;
+
+            let decryption_key =
+                BackupDecryptionKey::from_base64(maybe_recovery_key).map_err(|e| {
+                    <serde_json::Error as serde::de::Error>::custom(format!(
+                        "Couldn't deserialize the backup recovery key: {e:?}"
+                    ))
+                })?;
+
+            // Let's try to see if there's a backup on the homeserver.
+            let current_version = self.get_current_version().await?;
+
+            let Some(current_version) = current_version else {
+                warn!("Tried to enable backups, but no backup version was found on the server.");
+                return Ok(false);
+            };
+
+            Span::current().record("backup_version", &current_version.version);
+
+            let backup_info: RoomKeyBackupInfo = current_version.algorithm.deserialize_as()?;
+            let stored_keys = olm_machine.backup_machine().get_backup_keys().await?;
+
+            if stored_keys.backup_version.as_ref() == Some(&current_version.version)
+                && self.are_enabled().await
+            {
+                // If we already have a backup enabled which is using the currently active
+                // backup version, do nothing but tell the caller using the return value that
+                // backups are enabled.
+                Ok(true)
+            } else if decryption_key.backup_key_matches(&backup_info) {
+                info!(
+                    "We have found the correct backup recovery key, storing the backup recovery \
+                     key and enabling backups"
+                );
+
+                // We're enabling a new backup, reset the `backed_up` flags on the room keys and
+                // remove any key/version we might have.
+                olm_machine.backup_machine().disable_backup().await?;
```

Comment:
Can we store the ref to backup machine so we don't repeat the fetch from the olm machine each time?

Reply:
Sure: 18ab71e790f84a3094d8edca2d63da0b61e62768.